plugin/caffe/caffe_data_iter.cc

```diff
@@ -0,0 +1,264 @@
+  bool flat;
+
+  DMLC_DECLARE_PARAMETER(CaffeDataParam) {
+    DMLC_DECLARE_FIELD(prototxt).set_default("layer{}")
+      .describe("Caffe's layer parameter");
+    DMLC_DECLARE_FIELD(flat).set_default(false)
+      .describe("Augmentation Param: Whether to flat the data into 1D.");
+    DMLC_DECLARE_FIELD(num_examples).set_lower_bound(1).set_default(10000)
+      .describe("Number of examples in the epoch.");
+  }
+};
+
+struct CaffeDataIterWrapperParam : dmlc::Parameter<CaffeDataIterWrapperParam> {
+    /*! \brief data type */
+    int dtype;
+
+    DMLC_DECLARE_PARAMETER(CaffeDataIterWrapperParam) {
+      DMLC_DECLARE_FIELD(dtype)
+        .add_enum("float32", mshadow::kFloat32)
+        .add_enum("float64", mshadow::kFloat64)
+        .add_enum("float16", mshadow::kFloat16)
+        .set_default(mshadow::default_type_flag)
+        .describe("Data type.");
+    }
+};
+
+template<typename Dtype>
+class CaffeDataIter : public IIterator<TBlobBatch> {
+ public:
+  CaffeDataIter(void) : loc_(0), batch_size_(0), channels_(0), width_(0), height_(0) {}
+  virtual ~CaffeDataIter(void) {}
+
+  // intialize iterator loads data in
+  virtual void Init(const std::vector<std::pair<std::string, std::string> >& kwargs) {
+    std::map<std::string, std::string> kmap(kwargs.begin(), kwargs.end());
+    param_.InitAllowUnknown(kmap);
+
+    // Caffe seems to understand phase inside an "include {}" block
+    if (!param_.prototxt.has_phase()) {
+      if (param_.prototxt.include().size()) {
+        if (param_.prototxt.include(0).has_phase()) {
+          param_.prototxt.set_phase(param_.prototxt.include(0).phase());
+        }
+      }
+    }
+
+    std::string type = param_.prototxt.type();
+    caffe_data_layer_ = caffe::LayerRegistry<Dtype>::CreateLayer(param_.prototxt);
+    CHECK(caffe_data_layer_ != nullptr) << "Failed creating caffe data layer";
+    const size_t top_size = param_.prototxt.top_size();
+    if (top_size > 0) {
+      if(top_size > NR_SUPPORTED_TOP_ITEMS) {
+        LOG(WARNING) << "Too may \"top\" items, only two (one data, one label) are currently supported";
+      }
+      top_.reserve(top_size);
+      for (size_t x = 0; x < top_size; ++x) {
+        ::caffe::Blob<Dtype> *blob = new ::caffe::Blob<Dtype>();
+        cleanup_blobs_.push_back(std::unique_ptr<::caffe::Blob<Dtype>>(blob));
+        top_.push_back(blob);
+      }
+      caffe_data_layer_->SetUp(bottom_, top_);
+      const std::vector<int> &shape = top_[DATA]->shape();
+      const size_t shapeDimCount = shape.size();
+      if (shapeDimCount > 0) {
+        batch_size_ = shape[0];
+        if (shapeDimCount > 1) {
+          channels_ = shape[1];
+          if (shapeDimCount > 2) {
+            width_ = shape[2];
+            if (shapeDimCount > 3) {
+              height_ = shape[3];
+            }
+          }
+        }
+      }
+
+      if(top_size > DATA) {
+        if (param_.flat) {
+          batch_data_.shape_ = mshadow::Shape2(batch_size_, width_ * height_);
```

Comment:
You comment earlier:
"use mshadow::Shape2(batch_size_, width_ * height_) or mshadow::Shape4(batch_size_, width_ * height_, 1, 1).
I prefer the first"

No channel in your suggestion. I am confused.